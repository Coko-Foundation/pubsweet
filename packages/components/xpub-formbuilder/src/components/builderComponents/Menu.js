import React from 'react'
import { Menu } from '@pubsweet/ui'

const ValidationMenu = input => <Menu {...input} />

export default ValidationMenu
