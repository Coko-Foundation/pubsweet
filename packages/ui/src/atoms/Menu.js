import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import th from '../helpers/themeHelper'

// TODO: match the width of the container to the width of the widest option?
// TODO: use a <select> element instead of divs?
// TODO: determine the correct color for non-selected options
// FIXME: putting markup inside a <button> is invalid

// #region styled components
const Root = styled.div`
  max-width: calc(${th('gridUnit')} * 14);
  margin-bottom: ${props => (props.inline ? '0' : props.theme.gridUnit)};
`

const Label = styled.label`
  font-size: ${th('fontSizeBaseSmall')};
  display: block;
`

const Opener = styled.button.attrs({
  type: 'button',
})`
  background: transparent;
  border: ${th('borderWidth')} ${th('borderStyle')} ${th('colorBorder')};
  border-radius: ${th('borderRadius')};
  cursor: pointer;
  font-family: inherit;

  width: 100%;
  height: calc(${th('gridUnit')} * 2);
  padding: 0;

  display: flex;
  align-items: center;

  &:hover {
    border-color: ${th('colorPrimary')};
  }
`

const Value = styled.span`
  flex-grow: 1;

  text-align: left;
  padding: 0 calc(${th('gridUnit')} / 2);

  &:hover {
    color: ${th('colorPrimary')};
  }
`

const Placeholder = Value.extend`
  color: ${th('colorTextPlaceholder')};
  font-style: italic;
`

const ArrowContainer = styled.span`
  border-left: ${th('borderWidth')} ${th('borderStyle')} ${th('colorFurniture')};

  width: calc(${th('gridUnit')} * 2);
  height: calc(${th('gridUnit')} * 2 - ${th('borderWidth')} * 2);

  display: flex;
  align-items: center;
  justify-content: center;
`

const Arrow = styled.span`
  font-size: 50%;
  transition: transform 0.2s;
  transform: scaleX(2) scaleY(${props => (props.open ? -1.2 : 1.2)});
`

const Main = styled.div.attrs({
  role: 'listbox',
})`
  position: relative;
`

const OptionsContainer = styled.div`
  position: absolute;
  left: 0;
  right: 0;
`

const Options = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;

  background-color: ${th('colorBackground')};
  border: ${th('borderWidth')} ${th('borderStyle')} ${th('colorBorder')};
  border-radius: ${th('borderRadius')};
  overflow: hidden;
`

const Option = styled.div.attrs({
  role: 'option',
  tabIndex: '0',
  'aria-selected': props => props.active,
})`
  color: ${props => (props.active ? props.theme.textColor : '#444')};
  font-weight: ${props => (props.active ? '600' : 'inherit')};
  cursor: pointer;
  font-family: ${th('fontAuthor')};
  padding: calc(${th('subGridUnit')} - ${th('borderWidth')} * 2)
    calc(${th('subGridUnit')} * 2);
  border: ${th('borderWidth')} ${th('borderStyle')} transparent;
  border-width: ${th('borderWidth')} 0 ${th('borderWidth')} 0;
  white-space: nowrap;

  &:hover {
    background: ${th('colorBackgroundHue')};
    border-color: ${th('colorBorder')};
  }

  &:first-child:hover {
    border-top-color: ${th('colorBackgroundHue')};
  }

  &:last-child:hover {
    border-bottom-color: ${th('colorBackgroundHue')};
  }
`
// #endregion

class Menu extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      open: false,
      selected: props.value,
    }
  }

  toggleMenu = () => {
    this.setState({
      open: !this.state.open,
    })
  }

  handleSelect = selected => {
    this.setState({
      open: false,
      selected,
    })

    this.props.onChange(selected)
  }

  handleKeyPress = (event, selected) => {
    if (event.which === 13) {
      this.handleSelect(selected)
    }
  }

  optionLabel = value => {
    const { options } = this.props

    return options.find(option => option.value === value).label
  }

  render() {
    const {
      label,
      options,
      inline,
      placeholder,
      renderOption: RenderOption,
    } = this.props
    const { open, selected } = this.state

    return (
      <Root inline={inline} open={open}>
        {label && <Label>{label}</Label>}

        <Main>
          <Opener onClick={this.toggleMenu} open={open}>
            {selected ? (
              <Value>{this.optionLabel(selected)}</Value>
            ) : (
              <Placeholder>{placeholder}</Placeholder>
            )}
            <ArrowContainer>
              <Arrow open={open}>â–¼</Arrow>
            </ArrowContainer>
          </Opener>

          <OptionsContainer>
            {open && (
              <Options open={open}>
                {options.map(option => (
                  <RenderOption
                    handleKeyPress={this.handleKeyPress}
                    handleSelect={this.handleSelect}
                    key={option.value}
                    label={option.label}
                    selected={selected}
                    value={option.value}
                  />
                ))}
              </Options>
            )}
          </OptionsContainer>
        </Main>
      </Root>
    )
  }
}

const MenuOption = ({
  selected,
  label,
  value,
  handleSelect,
  handleKeyPress,
}) => (
  <Option
    active={value === selected}
    key={value}
    onClick={() => handleSelect(value)}
    onKeyPress={event => handleKeyPress(event, value)}
  >
    {label || value}
  </Option>
)

Menu.propTypes = {
  /** Menu items. */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.any.isRequired,
    }),
  ).isRequired,
  /** Custom option component. The component will be rendered with *optionProps*. */
  renderOption: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
  /** Optional label to be shown above the menu. */
  label: PropTypes.string,
  /** Placeholder until a value is selected. */
  placeholder: PropTypes.string,
}

Menu.defaultProps = {
  renderOption: MenuOption,
  placeholder: 'Choose in the list',
}

export default Menu
