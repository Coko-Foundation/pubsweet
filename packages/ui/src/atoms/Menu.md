### Option props

All menu item options will be passed the same props.

* _selected_ - the current selected option
* _label_ - the label of the option
* _value_ - the value of the option
* _handleSelect_ - handler called when clicking on the option
* _handleKeyPress_ - keyboard handler

A menu for selecting one of a list of options.

```js
const options = [
  { value: 'foo', label: 'Foo' },
  { value: 'bar', label: 'Bar' },
  { value: 'baz', label: 'Baz' },
]
;<Menu options={options} onChange={value => console.log(value)} />
```

When an option is selected, it replaces the placeholder.

```js
const options = [
  { value: 'foo', label: 'Foo' },
  { value: 'bar', label: 'Bar' },
  { value: 'baz', label: 'Baz' },
]
;<Menu options={options} value="foo" onChange={value => console.log(value)} />
```

A menu can have a label

```js
const options = [
  { value: 'foo', label: 'Foo' },
  { value: 'bar', label: 'Bar' },
  { value: 'baz', label: 'Baz' },
]
;<Menu
  options={options}
  label="Title"
  value="foo"
  onChange={value => console.log(value)}
/>
```

A menu can use a custom option component

```js
const CustomOption = ({ selected, label, value, handleSelect }) => (
  <div onClick={() => handleSelect(value)}>
    <span>Custom thing: {label}</span>
  </div>
)

const options = [
  { value: 'foo', label: 'Foo' },
  { value: 'bar', label: 'Bar' },
  { value: 'baz', label: 'Baz' },
]
;<Menu
  options={options}
  label="Title"
  value="foo"
  onChange={value => console.log(value)}
  renderOption={CustomOption}
/>
```
