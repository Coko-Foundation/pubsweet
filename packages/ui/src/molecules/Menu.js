import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { th, override } from '@pubsweet/ui-toolkit'

// TODO: match the width of the container to the width of the widest option?
// TODO: use a <select> element instead of divs?
// TODO: determine the correct color for non-selected options
// TODO: add z-indexes in theme object?
// FIXME: putting markup inside a <button> is invalid

// #region styled components
const Root = styled.div``

const CloseOverlay = styled.div`
  background-color: transparent;
  position: fixed;
  bottom: 0;
  left: 0;
  top: 0;
  right: 0;
  z-index: 10;

  ${override('ui.Menu.CloseOverlay')};
`

const Label = styled.label`
  font-size: ${th('fontSizeBaseSmall')};
  display: block;

  ${override('ui.Menu.Label')};
`

const Opener = styled.button.attrs({
  type: 'button',
})`
  background: transparent;
  border: ${th('borderWidth')} ${th('borderStyle')} ${th('colorBorder')};
  border-radius: ${th('borderRadius')};
  cursor: pointer;
  font-family: inherit;

  width: 100%;
  height: calc(${th('gridUnit')} * 6);
  padding: 0;

  display: flex;
  align-items: center;

  &:hover {
    border-color: ${th('colorPrimary')};
  }

  ${override('ui.Menu.Opener')};
`

const Value = styled.span`
  flex-grow: 1;

  text-align: left;
  padding: 0 ${th('gridUnit')};

  &:hover {
    color: ${th('colorPrimary')};
  }

  ${override('ui.Menu.Value')};
`

const Placeholder = Value.extend`
  color: ${th('colorTextPlaceholder')};
  font-style: italic;

  ${override('ui.Menu.Placeholder')};
`

const ArrowContainer = styled.span`
  border-left: ${th('borderWidth')} ${th('borderStyle')} ${th('colorFurniture')};

  width: calc(${th('gridUnit')} * 6);
  height: calc(${th('gridUnit')} * 6 - ${th('borderWidth')} * 2);

  display: flex;
  align-items: center;
  justify-content: center;

  ${override('ui.Menu.ArrowContainer')};
`

const Arrow = styled.span`
  font-size: 50%;
  transition: transform 0.2s;
  transform: scaleX(2) scaleY(${props => (props.open ? -1.2 : 1.2)});

  ${override('ui.Menu.Arrow')};
`

const Main = styled.div.attrs({
  role: 'listbox',
})`
  position: relative;

  ${override('ui.Menu.Main')};
`

const OptionsContainer = styled.div`
  position: absolute;
  left: 0;
  right: 0;

  ${override('ui.Menu.OptionsContainer')};
`

const Options = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;

  background-color: ${th('colorBackground')};
  border: ${th('borderWidth')} ${th('borderStyle')} ${th('colorBorder')};
  border-radius: ${th('borderRadius')};
  overflow-y: auto;
  max-height: ${({ maxHeight }) => `${maxHeight}px`};
  z-index: 100;

  ${override('ui.Menu.Options')};
`

const Option = styled.div.attrs({
  role: 'option',
  tabIndex: '0',
  'aria-selected': props => props.active,
})`
  color: ${props => (props.active ? props.theme.textColor : '#444')};
  font-weight: ${props => (props.active ? '600' : 'inherit')};
  cursor: pointer;
  font-family: ${th('fontAuthor')};
  padding: calc(${th('gridUnit')} - ${th('borderWidth')} * 2)
    calc(${th('gridUnit')} * 2);
  border: ${th('borderWidth')} ${th('borderStyle')} transparent;
  border-width: ${th('borderWidth')} 0 ${th('borderWidth')} 0;
  white-space: nowrap;

  &:hover {
    background: ${th('colorBackgroundHue')};
    border-color: ${th('colorBorder')};
  }

  &:first-child:hover {
    border-top-color: ${th('colorBackgroundHue')};
  }

  &:last-child:hover {
    border-bottom-color: ${th('colorBackgroundHue')};
  }

  ${override('ui.Menu.Option')};
`
// #endregion

class Menu extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      open: false,
      selected: props.value,
    }
  }

  toggleMenu = () => {
    this.setState({
      open: !this.state.open,
    })
  }

  handleSelect = ({ selected, open }) => {
    this.setState({
      open,
      selected,
    })
    if (this.props.onChange) this.props.onChange(selected)
  }

  handleKeyPress = (event, selected, open) => {
    if (event.which === 13) {
      this.handleSelect(selected, open)
    }
  }

  optionLabel = value => {
    const { options } = this.props

    return options.find(option => option.value === value).label
  }

  render() {
    const {
      maxHeight = 250,
      label,
      options,
      inline,
      placeholder,
      renderOption: RenderOption,
      renderOpener: RenderOpener,
      className,
    } = this.props
    const { open, selected } = this.state

    return (
      <Root className={className} inline={inline} open={open}>
        {label && <Label>{label}</Label>}
        {open && <CloseOverlay onClick={this.toggleMenu} />}
        <Main>
          <RenderOpener
            open={open}
            optionLabel={this.optionLabel}
            placeholder={placeholder}
            selected={selected}
            toggleMenu={this.toggleMenu}
          />
          <OptionsContainer>
            {open && (
              <Options maxHeight={maxHeight} open={open}>
                {options.map(option => (
                  <RenderOption
                    handleKeyPress={this.handleKeyPress}
                    handleSelect={this.handleSelect}
                    key={option.value}
                    label={option.label}
                    selected={selected}
                    value={option.value}
                  />
                ))}
              </Options>
            )}
          </OptionsContainer>
        </Main>
      </Root>
    )
  }
}

const DefaultMenuOption = ({
  selected,
  label,
  value,
  handleSelect,
  handleKeyPress,
}) => (
  <Option
    active={value === selected}
    key={value}
    onClick={() => handleSelect({ open: false, selected: value })}
    onKeyPress={event => handleKeyPress(event, value)}
  >
    {label || value}
  </Option>
)

const DefaultOpener = ({
  toggleMenu,
  open,
  selected,
  placeholder,
  optionLabel,
}) => (
  <Opener onClick={toggleMenu} open={open}>
    {selected ? (
      <Value>{optionLabel(selected)}</Value>
    ) : (
      <Placeholder>{placeholder}</Placeholder>
    )}
    <ArrowContainer>
      <Arrow open={open}>â–¼</Arrow>
    </ArrowContainer>
  </Opener>
)

Menu.propTypes = {
  /** Menu items. */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.any.isRequired,
    }),
  ).isRequired,
  /** Custom component for the selected option. */
  renderOpener: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
  /** Custom option component. The component will be rendered with *optionProps*. */
  renderOption: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
  /** Optional label to be shown above the menu. */
  label: PropTypes.string,
  /** Placeholder until a value is selected. */
  placeholder: PropTypes.string,
  /** Maximum height of the options container. */
  maxHeight: PropTypes.number,
}

Menu.defaultProps = {
  renderOption: DefaultMenuOption,
  renderOpener: DefaultOpener,
  placeholder: 'Choose in the list',
}

export default Menu
