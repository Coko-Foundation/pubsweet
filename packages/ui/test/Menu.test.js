import React from 'react'
import renderer from 'react-test-renderer'

import Menu from '../src/atoms/Menu'

const props = {
  options: [{ label: 'Foo', value: 'foo' }, { label: 'Bar', value: 'bar' }],
  value: 'foo',
}

describe('Menu', () => {
  test('is rendered correctly', () => {
    const tree = renderer.create(<Menu {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
