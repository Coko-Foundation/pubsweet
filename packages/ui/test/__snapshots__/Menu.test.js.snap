// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu is rendered correctly 1`] = `
.c1 {
  background: transparent;
  border: 1px solid grey;
  border-radius: 8px;
  cursor: pointer;
  font-family: inherit;
  width: 100%;
  height: calc(8px * 6);
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:hover {
  border-color: black;
}

.c2 {
  border-right: 1px solid grey;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  white-space: nowrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  line-height: 8px;
  padding: 0 8px;
  white-space: nowrap;
}

.c2:hover {
  color: black;
}

.c3 {
  width: calc(8px * 2);
  height: calc(8px * 2 - 1px * 2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: calc(8px * 2);
}

.c4 {
  font-size: 50%;
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: scaleX(2) scaleY(1.2);
  -ms-transform: scaleX(2) scaleY(1.2);
  transform: scaleX(2) scaleY(1.2);
}

.c0 {
  position: relative;
}

.c5 {
  position: absolute;
  left: 0;
  right: 0;
}

<div
  className=""
  open={false}
>
  <div
    className="c0"
    role="listbox"
  >
    <button
      className="c1"
      onClick={[Function]}
      open={false}
      type="button"
    >
      <span
        className="c2"
      >
        Foo
      </span>
      <span
        className="c3"
      >
        <span
          className="c4"
          open={false}
        >
          ▼
        </span>
      </span>
    </button>
    <div
      className="c5"
    />
  </div>
</div>
`;

exports[`Menu is rendered correctly when open 1`] = `
.c0 {
  background-color: transparent;
  position: fixed;
  bottom: 0;
  left: 0;
  top: 0;
  right: 0;
  z-index: 10;
}

.c2 {
  background: transparent;
  border: 1px solid grey;
  border-radius: 8px;
  cursor: pointer;
  font-family: inherit;
  width: 100%;
  height: calc(8px * 6);
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2:hover {
  border-color: black;
}

.c3 {
  border-right: 1px solid grey;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  white-space: nowrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  line-height: 8px;
  padding: 0 8px;
  white-space: nowrap;
}

.c3:hover {
  color: black;
}

.c4 {
  width: calc(8px * 2);
  height: calc(8px * 2 - 1px * 2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: calc(8px * 2);
}

.c5 {
  font-size: 50%;
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: scaleX(2) scaleY(-1.2);
  -ms-transform: scaleX(2) scaleY(-1.2);
  transform: scaleX(2) scaleY(-1.2);
}

.c1 {
  position: relative;
}

.c6 {
  position: absolute;
  left: 0;
  right: 0;
}

.c7 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  background-color: white;
  border: 1px solid grey;
  border-radius: 8px;
  overflow-y: auto;
  max-height: 250px;
  z-index: 100;
}

.c8 {
  font-weight: 600;
  cursor: pointer;
  padding: calc(8px - 1px * 2) calc(8px * 2);
  border: 1px solid transparent;
  border-width: 1px 0 1px 0;
  white-space: nowrap;
}

.c8:hover {
  background: grey;
  border-color: grey;
}

.c8:first-child:hover {
  border-top-color: grey;
}

.c8:last-child:hover {
  border-bottom-color: grey;
}

.c9 {
  color: #444;
  font-weight: inherit;
  cursor: pointer;
  padding: calc(8px - 1px * 2) calc(8px * 2);
  border: 1px solid transparent;
  border-width: 1px 0 1px 0;
  white-space: nowrap;
}

.c9:hover {
  background: grey;
  border-color: grey;
}

.c9:first-child:hover {
  border-top-color: grey;
}

.c9:last-child:hover {
  border-bottom-color: grey;
}

<div
  className=""
  open={true}
>
  <div
    className="c0"
    onClick={[Function]}
  />
  <div
    className="c1"
    role="listbox"
  >
    <button
      className="c2"
      onClick={[Function]}
      open={true}
      type="button"
    >
      <span
        className="c3"
      >
        Foo
      </span>
      <span
        className="c4"
      >
        <span
          className="c5"
          open={true}
        >
          ▼
        </span>
      </span>
    </button>
    <div
      className="c6"
    >
      <div
        className="c7"
        open={true}
      >
        <div
          aria-selected={true}
          className="c8"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="option"
          tabIndex="0"
        >
          Foo
        </div>
        <div
          aria-selected={false}
          className="c9"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="option"
          tabIndex="0"
        >
          Bar
        </div>
      </div>
    </div>
  </div>
</div>
`;
